### PR TITLE
Jenkins postbuild plugin

### PR DESCRIPTION
**Pull request by pritic**
Bio: I am Priti Changlani, a summer intern at Rackspace US, Inc. and a Computer Science graduate student at University of Florida. 
author(s) social media contact info: https://www.linkedin.com/in/pritichanglani
accept or decline to allow redaction, editing or structure changes by technical writer staff: Accept
desired date of publishing: 08/21/2015

### CODE REVIEW

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+title: "Jenkins Post-build Plugin"
+date: 2015-08-21 14:00
+comments: true
+author: Priti Changlani
+published: true
+categories:
+    - jenkins
+    - plugin
+---
+This is the first part of a two part tutorial series to develop and Jenkins plugin
```
Comment (annegentle) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: two-part

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+ test-suites to cover the entire API contract. In the process, I found many resources were
+ hard to follow and some were quite out of date. I
+ realized the need for
+ an updated, easy and dedicated tutorial on this topic and therefore decided to write one
+ myself!
+
+This part of the
+tutorial talks about Jenkins plugins as a
+whole - what projects are
+they, how they are created, what do they depend on, how do you run them, et al. We will
```
Comment (annegentle) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: rather than "et al" how about "and so on." It's a little more colloquial and friendly for the blog tone.

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+ concentrate more on the technical aspects of the plugins so that next time you have an
+  extension idea, you don't wait on Jenkins and write a plugin yourself! So
+let's get started!
```
Comment (annegentle) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: Use exclamation points sparingly -- so far you are on your fourth. Pick one place to sound so! Enthusiastic! :)

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+1. [More with Maven](#more-with-maven)
+1. [Introduction to Part 2](#intro-to-part-2)
+
+## <a name="setting-up"></a>Setting Up
+
+Jenkins plugins are basically [Maven](https://maven.apache.org/) projects
+with Java modules and therefore the first and foremost requirement is to
+have compatible JDK and Maven versions.
+
+For this tutorial, we are going to need the following installations:
```
Comment (annegentle) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: we need (more concise, easier to read)

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+
+JDK: [jdk1.7.0_79.jdk](http://www.oracle.com/technetwork/java/javase/downloads/jre7-downloads-1880261.html)
+
+Maven: [apache-maven-3.3.3](https://maven.apache.org/download.cgi)
+
+Jenkins: [Installation](https://wiki.jenkins-ci.org/display/JENKINS/Installing+Jenkins)
+
+IDE: [IntelliJ IDEA 14.1.4](https://www.jetbrains.com/idea/download/)
+
+The next thing would be to configure maven for the user. Navigate to your
```
Comment (annegentle) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: Maven

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+`${user.home}/.m2/settings.xml` and add the following <localRepository> tag to
+ the `<settings>` block
+
+    <localRepository>/path/to/local/repo/</localRepository>
+
+ This path is the directory where all the maven dependencies would be
```
Comment (annegentle) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: Maven

File: _posts/2015-08-21-jenkins-post-build-plugin.md (hunk 3 above)
Comment (pritic) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: Haha, I'll bring the number down :)

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+            </pluginRepository>
+          </pluginRepositories>
+        </profile>
+      </profiles>
+    </settings>
+
+
+## <a name="creating-the-plugin"></a>Creating the Plugin
+
+To start with the plugin source code creation, in the terminal type
```
Comment (annegentle) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: Let's stick with colons consistently before examples.

, in the terminal, type:

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+
+    mvn hpi:create -Pjenkins
+
+The changes in the `settings.xml` enable us to use this short command instead of the
+below actual command
```
Comment (annegentle) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: this short command:

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+
+    mvn org.jenkins-ci.tools:maven-hpi-plugin:<plugin-version>:create
+
+You would be prompted to enter a groupId. We would select the default
```
Comment (annegentle) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: default trails off here, so I'm not sure what you mean?

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+
+    Enter the groupId of your plugin [org.jenkins-ci.plugins]: org.jenkinsci.plugins
+
+You then need to provide an artifactId for the plugin. This id becomes the
+package name for the project and it is a good practice to keep the package
+name same as the plugin name (that the end users would see). The suffix
+'-plugin' is unnecessary since it is implied that this is a Jenkins plugin.
+For the example here, we would name the plugin 'testExample'.
+
+    Enter the artifactId of your plugin (normally without '-plugin' suffix): testExample
```
Comment (annegentle) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: not sure how this indent will render?

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+was executed.
+
+## <a name="understanding-the-project-structure"></a>Understanding the Project Structure
+
+<img class="blog-post right" src="{% asset_path 2015-08-21-jenkins-post-build-plugin/project_structure.png %}"/>
+Opening the project in IntelliJ IDEA should display the project structure in the left
+pane.
+
+Notice how the `groupId` is appended with the `artifactId` and forms a directory under the
+`java` directory. This is the location where the java code resides. Creating a project,
```
Comment (annegentle) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: Java code here because it needs to be the proper name unlike `java` which can remain as-is.

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+files under
+`resources/org/jenkinsci/plugins/testExample/HelloWorldBuilder`
+to see which part of the UI they render.
+`resources/index.jelly` renders the view on the 'Installed Plugins' page under 'Manage
+Plugins'.
+
+## <a name="debugging"></a>Debugging the Plugin
+
+In IntelliJ IDEA, navigate to Run>Edit Configurations>Remote and set the Port to 8000,
+which is the default mvnDebug port. Hit 'Apply'
```
Comment (annegentle) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: Click rather than Hit -- hit feels a little violent :)

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+![Debugger Configuration]({% asset_path 2015-08-21-jenkins-post-build-plugin/debug_config.png %})
+
+In the terminal, type the following command:
+
+    user@localhost:~/testExample$ mvnDebug hpi:run
+    Preparing to Execute Maven in Debug Mode
+    Listening for transport dt_socket at address: 8000
+
+This starts the listener on port 8000. As the message states, maven is ready in
```
Comment (annegentle) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: Maven

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+
+    user@localhost:~/testExample$ mvn install
+    .
+    . <wait for the processing>
+    .
+    [INFO] ------------------------------------------------------------------------
+    [INFO] BUILD SUCCESS
+    [INFO] ------------------------------------------------------------------------
+
+The install command installs all the maven dependencies (if not installed already)
```
Comment (annegentle) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: Maven

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 15]
```diff
@@ -0,0 +1,241 @@
...
+    .
+    .<wait for the processing>
+    .
+    INFO: Jenkins is fully up and running
+
+When you see the above info message, type
+http://localhost:8080/jenkins/pluginManager/installed in your browser and notice that
+the plugin is now present in the installed plugin list. This is where the debugger,
+running in the background,
+performed its magic!
```
Comment (annegentle) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: You know, I am good with this exclamation point now! :)

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 16]
```diff
@@ -0,0 +1,241 @@
...
+
+<img class="blog-post center" src="{% asset_path 2015-08-21-jenkins-post-build-plugin/plugin_installed.png %}"/>
+
+<img class="blog-post right" src="{% asset_path 2015-08-21-jenkins-post-build-plugin/maven-projects.png %}"/>
+
+## <a name="more-with-maven"></a>More with Maven
+
+A developer can do much more with Maven and doesn't have to remember all the commands.
+Clicking on 'Maven Projects' on the right side panel expands all the available options
+for the developer. Double clicking on any executes the command.
```
Comment (annegentle) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: Double-clicking on any project option?

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 17]
```diff
@@ -0,0 +1,241 @@
...
+
+While a lifecycle state can be executed with the command
```
Comment (annegentle) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: colon

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 18]
```diff
@@ -0,0 +1,241 @@
...
+
+    mvn <state>
+
+a maven plugin can be used as
```
Comment (annegentle) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: Maven and colon after as

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 19]
```diff
@@ -0,0 +1,241 @@
...
+
+## <a name="intro-to-part-2"></a>Introduction to Part 2
+
+The auto-generated testExample project is an example of a Jenkins build plugin and you may
+ see
+that the class `HelloWorldBuilder.java` extends the class `hudson.tasks.Builder`, however,
+in the upcoming part 2 of this tutorial, we will create a post-build plugin and therefore
+extend the class `hudson.tasks.Recorder` and make the `Extension` class
+`DescriptorImpl` extend the class `BuildStepDescriptor<Publisher>` instead of
+`BuildStepDescriptor<Builder>`
```
Comment (annegentle) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: Period missing.

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 20]
```diff
@@ -0,0 +1,243 @@
...
+title: "Jenkins Post-build Plugin"
+date: 2015-08-21 14:00
+comments: true
+author: Priti Changlani
+published: true
+categories:
+    - jenkins
+    - plugin
+---
+This is the first part of a two-part tutorial series to develop and Jenkins plugin
```
Comment (catlook) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: ... to develop a Jenkins plugin ,,,

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 21]
```diff
@@ -0,0 +1,243 @@
...
+(specifically, Jenkins post-build plugin). Jenkins is a very popular continuous
```
Comment (catlook) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: ...  very popular continuous-integration ...

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 22]
```diff
@@ -0,0 +1,243 @@
...
+integration tool and with the small amount of (scattered) information present
```
Comment (catlook) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: tool and with the small ...

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 23]
```diff
@@ -0,0 +1,243 @@
...
+ out there, it is really hard for a beginner to dive in this amazing area of extending
```
Comment (catlook) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: ... there. It is really ...

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 24]
```diff
@@ -0,0 +1,243 @@
...
+ it - Plugins!
+
+<!-- more -->
+
+I am [Priti Changlani](https://www.linkedin.com/in/pritichanglani), a summer intern at
+Rackspace US, Inc. and a Computer Science graduate student at
+University of Florida. At Rackspace, I have worked as a Quality Engineer in teams like
+[Solum](https://github.com/stackforge/solum), [Repose](https://github.com/rackerlabs/repose)
+and [Containers](https://wiki.openstack.org/wiki/ContainersTeam). A part of my job was to enhance the process of quality engineering
```
Comment (catlook) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: ... As part of my job, I enhanced ...

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 25]
```diff
@@ -0,0 +1,243 @@
...
+as a
+ whole and to that effect, I developed a Jenkins plugin - [API Coverage]
+ (https://github.com/rackerlabs/api-coverage) for QEs to
+ measure the
+ ability of their
+ test-suites to cover the entire API contract. In the process, I found
+ many resources were
+ hard to follow and some were quite out of date. I
```
Comment (catlook) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: ... follow, and some ...

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 26]
```diff
@@ -0,0 +1,243 @@
...
+ realized the need for
+ an updated, easy and dedicated tutorial on this topic and therefore decided to write one
```
Comment (catlook) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: ...  easy, and dedicated ...

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 27]
```diff
@@ -0,0 +1,243 @@
...
+ myself.
+
+This part of the
+tutorial talks about Jenkins plugins as a
+whole - what projects are
+they, how they are created, what do they depend on, how do you run them and so on. We will
+ concentrate more on the technical aspects of the plugins so that next time you have an
+  extension idea, you don't wait on Jenkins and write a plugin yourself.
```
Comment (catlook) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: ... you won't wait on Jenkins but, instead, write a plugin yourself.

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 28]
```diff
@@ -0,0 +1,243 @@
...
+
+Let's get started...
```
Comment (catlook) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: Let's get started.

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 29]
```diff
@@ -0,0 +1,243 @@
...
+1. [Creating the Plugin](#creating-the-plugin)
+1. [Understanding the Project Structure](#understanding-the-project-structure)
+1. [Debugging the plugin](#debugging)
+1. [More with Maven](#more-with-maven)
+1. [Introduction to Part 2](#intro-to-part-2)
+
+## <a name="setting-up"></a>Setting Up
+
+Jenkins plugins are basically [Maven](https://maven.apache.org/) projects
+with Java modules and therefore the first and foremost requirement is to
```
Comment (catlook) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: ... therefore the main requirement ...

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 30]
```diff
@@ -0,0 +1,243 @@
...
+
+JDK: [jdk1.7.0_79.jdk](http://www.oracle.com/technetwork/java/javase/downloads/jre7-downloads-1880261.html)
+
+Maven: [apache-maven-3.3.3](https://maven.apache.org/download.cgi)
+
+Jenkins: [Installation](https://wiki.jenkins-ci.org/display/JENKINS/Installing+Jenkins)
+
+IDE: [IntelliJ IDEA 14.1.4](https://www.jetbrains.com/idea/download/)
+
+The next thing would be to configure Maven for the user. Navigate to your
```
Comment (catlook) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: Next, we configure Maven ...

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 31]
```diff
@@ -0,0 +1,243 @@
...
+`${user.home}/.m2/settings.xml` and add the following <localRepository> tag to
+ the `<settings>` block.
+
+    <localRepository>/path/to/local/repo/</localRepository>
+
+ This path is the directory where all the Maven dependencies would be
```
Comment (catlook) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: ... Maven dependencies are ...

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 32]
```diff
@@ -0,0 +1,243 @@
...
+ downloaded and a good practice is to have the default value `${user.home}/.m2/repository/`.
```
Comment (catlook) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: ... downloaded. As a good practice, set the default ...

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 33]
```diff
@@ -0,0 +1,243 @@
...
+
+For developing a Jenkins plugin, developers need to have a `<pluginGroup>` and a
+jenkins `<profile>` in the `${user.home}/.m2/settings.xml`. Post these
```
Comment (catlook) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: ... the `${user.home}/.m2/settings.xml`. To post these ...

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 34]
```diff
@@ -0,0 +1,243 @@
...
+      </profiles>
+    </settings>
+
+The changes in the `settings.xml` also enable us to use a shorter command for creating the plugin in lieu of the longer version:
+
+    mvn org.jenkins-ci.tools:maven-hpi-plugin:<plugin-version>:create
+
+## <a name="creating-the-plugin"></a>Creating the Plugin
+
+To start with the plugin source code creation, in the terminal type:
```
Comment (catlook) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: ... creation, type in the terminal: ...

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 35]
```diff
@@ -0,0 +1,243 @@
...
+
+    mvn hpi:create
+
+You would be prompted to enter a groupId. We would select the default structure - `org.jenkinsci.plugins`
```
Comment (catlook) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: When prompted to enter a groupId, select the default structure - `org.jenkinsci.plugins` .

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 36]
```diff
@@ -0,0 +1,243 @@
...
+
+    Enter the groupId of your plugin [org.jenkins-ci.plugins]: org.jenkinsci.plugins
+
+You then need to provide an artifactId for the plugin. This id becomes the
```
Comment (catlook) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: Then provide an ...

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 37]
```diff
@@ -0,0 +1,243 @@
...
+package name for the project and it is a good practice to keep the package
```
Comment (catlook) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: ... project. As a good practice, keep ...

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 38]
```diff
@@ -0,0 +1,243 @@
...
+name same as the plugin name (that the end users would see). The suffix
+'-plugin' is unnecessary since it is implied that this is a Jenkins plugin.
```
Comment (catlook) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: ... unnecessary, because it ...

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 39]
```diff
@@ -0,0 +1,243 @@
...
+For the example here, we would name the plugin 'testExample'.
+
+    Enter the artifactId of your plugin (normally without '-plugin' suffix): testExample
+
+You should now have a directory 'testExample' at the path the above command
```
Comment (catlook) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: ... the path where the above ...

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 40]
```diff
@@ -0,0 +1,243 @@
...
+was executed.
+
+## <a name="understanding-the-project-structure"></a>Understanding the Project Structure
+
+<img class="blog-post right" src="{% asset_path 2015-08-21-jenkins-post-build-plugin/project_structure.png %}"/>
+Opening the project in IntelliJ IDEA should display the project structure in the left
+pane.
+
+Notice how the `groupId` is appended with the `artifactId` and forms a directory under the
+`java` directory. This is the location where the Java code resides. Creating a project,
```
Comment (catlook) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: ... directory. The Java code resides here. Creating ...

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 41]
```diff
@@ -0,0 +1,243 @@
...
+In the terminal, type the following command:
+
+    user@localhost:~/testExample$ mvnDebug hpi:run
+    Preparing to Execute Maven in Debug Mode
+    Listening for transport dt_socket at address: 8000
+
+This starts the listener on port 8000. As the message states, Maven is ready in
+Debug mode. This makes debugging pretty easy. How? We'll see this when we run the plugin.
+
+The next step is to install the plugin. Enter the following command in a different
```
Comment (catlook) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: Next, install ...

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 42]
```diff
@@ -0,0 +1,243 @@
...
+
+    user@localhost:~/testExample$ mvn install
+    .
+    . <wait for the processing>
+    .
+    [INFO] ------------------------------------------------------------------------
+    [INFO] BUILD SUCCESS
+    [INFO] ------------------------------------------------------------------------
+
+The install command installs all the Maven dependencies (if not installed already)
```
Comment (catlook) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: This command installs ...

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 43]
```diff
@@ -0,0 +1,243 @@
...
+specified in `pom.xml` and generates the `target` and the `work` directories. It also
+generates the `target/testExample.hpi` file which is a complete package for the plugin
+code. To debug, one can just import this file in Jenkins (Manage Jenkins>Manage
```
Comment (catlook) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: ... To debug, just import ...

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 44]
```diff
@@ -0,0 +1,243 @@
...
+running in the background,
+performed its magic!
+
+<img class="blog-post center" src="{% asset_path 2015-08-21-jenkins-post-build-plugin/plugin_installed.png %}"/>
+
+<img class="blog-post right" src="{% asset_path 2015-08-21-jenkins-post-build-plugin/maven-projects.png %}"/>
+
+## <a name="more-with-maven"></a>More with Maven
+
+A developer can do much more with Maven and doesn't have to remember all the commands.
```
Comment (catlook) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: ... Maven without having to remember ....

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 45]
```diff
@@ -0,0 +1,243 @@
...
+Clicking on 'Maven Projects' on the right side panel expands all the available options
+for the developer. Double-clicking on any of these options executes the command.
+
+While a lifecycle state can be executed with the command:
```
Comment (catlook) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: A lifecycle ....

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 46]
```diff
@@ -0,0 +1,243 @@
...
+
+    mvn <state>
+
+a Maven plugin can be used as:
+
+    mvn <plugin>:<action>
+
+Using the above one may just execute `mvn hpi:hpi` to generate the `.hpi` file
```
Comment (catlook) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: ... the above, you just ...

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 47]
```diff
@@ -0,0 +1,243 @@
...
+
+## <a name="intro-to-part-2"></a>Introduction to Part 2
+
+The auto-generated testExample project is an example of a Jenkins build plugin and you may
```
Comment (catlook) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: ... build plugin and you can ...

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 48]
```diff
@@ -0,0 +1,243 @@
...
+ see
+that the class `HelloWorldBuilder.java` extends the class `hudson.tasks.Builder`, however,
```
Comment (catlook) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: ... class `hudson.tasks.Builder`. However, ...

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 49]
```diff
@@ -0,0 +1,243 @@
...
+in the upcoming part 2 of this tutorial, we will create a post-build plugin and therefore
```
Comment (catlook) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: ...  and, therefore, ...

File: _posts/2015-08-21-jenkins-post-build-plugin.md [hunk 50]
```diff
@@ -0,0 +1,243 @@
...
+extend the class `hudson.tasks.Recorder` and make the `Extension` class
+`DescriptorImpl` extend the class `BuildStepDescriptor<Publisher>` instead of
+`BuildStepDescriptor<Builder>`.
+
```
Comment (catlook) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: [ Ed note:  Very nice article!! I look forward to seeing the next installment.  Please use any of my edits that you like and discard the rest! -- Cat ]

File: _posts/2015-08-21-jenkins-post-build-plugin.md (hunk 23 above)
Comment (pritic) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: I have accepted almost all of the suggested edits :) except for this one. The 'it is really..' part is in continuation with the '..and with the small amount..'. Hope this makes sense.

File: _posts/2015-08-21-jenkins-post-build-plugin.md (hunk 33 above)
Comment (pritic) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: Here, I meant 'Post these' as 'After these..' Makes sense?

File: _posts/2015-08-21-jenkins-post-build-plugin.md (hunk 22 above)
Comment (pritic) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: Err..this is a bit unclear to me

File: _posts/2015-08-21-jenkins-post-build-plugin.md (hunk 33 above)
Comment (catlook) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: Yes!!  I misinterpreted the intent.  Go ahead and use "After" to be clearer :)

-- Cat Lookabaugh, Sr. Information Developer
Rackspace Hosting - the #1 Managed Cloud Company
DevDocs on Developer Experience Team
Office: (210) 312-8193   Cell: (210) 364-1060
cat.lookabaugh@rackspace.com

We should not write so that it is possible for the reader to understand us, but so that it is impossible for him to misunderstand us. - Quintilian, rhetorician (c. 35-100)

Harmony + Learner + Responsibility + Context + Intellection + Arranger + Consistency + Relator + Input + Analytical

File: _posts/2015-08-21-jenkins-post-build-plugin.md (hunk 22 above)
Comment (catlook) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: I wanted you to hyphenate the adjective "continuous-integration".  So the sentence would look like:


Jenkins is a very popular continuous-integration tool and with the small amount of (scattered) information present ...

Better?

-- Cat Lookabaugh, Sr. Information Developer
Rackspace Hosting - the #1 Managed Cloud Company
DevDocs on Developer Experience Team
Office: (210) 312-8193   Cell: (210) 364-1060
cat.lookabaugh@rackspace.com

We should not write so that it is possible for the reader to understand us, but so that it is impossible for him to misunderstand us. - Quintilian, rhetorician (c. 35-100)

Harmony + Learner + Responsibility + Context + Intellection + Arranger + Consistency + Relator + Input + Analytical

Comment (pritic) [inline _posts/2015-08-21-jenkins-post-build-plugin.md]: All righty! That's done :)

**Comment by pritic** [comment]
I have accepted all of the suggested edits. There are no more updates from my side. Please go ahead and merge it, if we are good with the final post. Thank you for all the help with the editing, @catlook @annegentle . I appreciate it :)

**Comment by pritic** [comment]
I was just wondering if you guys are waiting on something to merge this?
